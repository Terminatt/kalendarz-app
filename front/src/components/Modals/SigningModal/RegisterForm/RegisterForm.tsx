import React, { useState } from 'react';
import { Form, Input } from 'antd';
import CustomForm from '@components/CustomForm/CustomForm';
import FormUtils from '@utils/form';
import { useDispatch, useSelector } from 'react-redux';
import { RootState } from '@store/index';
import { registerAccount } from '@store/user/asyncActions';
import { UserRegisterErrorResponse } from '@store/user/types';

const { useForm } = Form;

export interface RegisterFormValues {
  first_name: string;
  last_name: string;
  username: string;
  email: string;
  password: string;
  repeat_password?: string;
  title: string;
}

export interface RegisterFormProps {
    onFinishCallback?: () => void;
}

const RegisterForm: React.FC<RegisterFormProps> = (props) => {
    const { onFinishCallback } = props;
    const [form] = useForm<RegisterFormValues>();
    const { isLoading } = useSelector((state: RootState) => state.user);
    const dispatch = useDispatch();
    const [errorResponse, setErrorResponse] = useState<null | UserRegisterErrorResponse>(null);

    const onFinish = (values: RegisterFormValues) => {
        const payload = { ...values };
        delete payload.repeat_password;

        dispatch(registerAccount({
            requestPayload: payload,
            onSuccess: () => {
                if (onFinishCallback) {
                    onFinishCallback();
                }
            },
            onError: (errorData) => {
                if (!errorData.response) {
                    return;
                }
                setErrorResponse(errorData.response.data);
            },
        }));
    };

    return (
        <CustomForm formProps={{ form, onFinish }} errorResponse={errorResponse} isLoading={isLoading} primaryBtnText="Zarejestruj się">
            <Form.Item
                className="half-input"
                label="Imię"
                name="first_name"
                rules={[FormUtils.getRequiredRule(), FormUtils.getMaxCharRule(24, 'Imię może mieć maksymalnie 24 znaki')]}
            >
                <Input placeholder="Podaj imię" />
            </Form.Item>
            <Form.Item
                className="half-input"
                label="Nazwisko"
                name="last_name"
                rules={[FormUtils.getRequiredRule(), FormUtils.getMaxCharRule(24, 'Nazwisko może mieć maksymalnie 24 znaki')]}
            >
                <Input placeholder="Podaj nazwisko" />
            </Form.Item>
            <Form.Item
                label="Nazwa użytkownika"
                name="username"
                rules={[FormUtils.getRequiredRule(), FormUtils.getMaxCharRule(24, 'Nazwa użytkownika może mieć maksymalnie 24 znaki')]}
            >
                <Input placeholder="Podaj nazwę użytkownika" />
            </Form.Item>
            <Form.Item
                label="Email"
                name="email"
                rules={[
                    FormUtils.getRequiredRule(),
                    FormUtils.getMaxCharRule(50, 'Adres email może mieć maksymalnie 50 znaków'),
                    FormUtils.getEmailRule('To nie jest poprawny adres email'),
                ]}
            >
                <Input placeholder="Podaj email" />
            </Form.Item>
            <Form.Item
                label="Hasło"
                name="password"
                rules={[
                    FormUtils.getRequiredRule(),
                    FormUtils.getMaxCharRule(24, 'Hasło może mieć maksymalnie 24 znaki'),
                    FormUtils.getMinCharRule(9, 'Hasło musi mieć conajmniej 9 znaków'),
                    FormUtils.getNumericPasswordRule(),
                    FormUtils.getSpecialCharacterPasswordRule(),
                ]}
            >
                <Input type="password" placeholder="Podaj hasło" />
            </Form.Item>
            <Form.Item
                dependencies={['password']}
                label="Powtórz hasło"
                name="repeat_password"
                rules={[FormUtils.getRequiredRule(), FormUtils.getRepeatPasswordRule('password')]}
            >
                <Input type="password" placeholder="Powtórz hasło" />
            </Form.Item>
            <Form.Item
                label="Tytuł"
                name="title"
                wrapperCol={{ span: 6 }}
                rules={[FormUtils.getMaxCharRule(24, 'Tytuł może mieć maksymalnie 24 znaki')]}
            >
                <Input placeholder="Podaj tytuł" />
            </Form.Item>
        </CustomForm>
    );
};

export default RegisterForm;
