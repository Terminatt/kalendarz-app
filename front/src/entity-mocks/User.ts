import { Group, User } from '@store/user/types';
import dayjs from 'dayjs';

export const userMock: User = {
    id: 1,
    firstName: 'Test',
    lastName: 'User',
    created: dayjs().toISOString(),
    email: 'test@gmail.com',
    username: 'Test User',
    groups: Group.REGULAR_USER,
};

export const adminUserMock: User = {
    id: 2,
    firstName: 'Test',
    lastName: 'User',
    created: dayjs().toISOString(),
    email: 'test@gmail.com',
    username: 'Test User',
    groups: Group.ADMIN,
};
