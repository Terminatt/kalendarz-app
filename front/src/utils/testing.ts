export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
