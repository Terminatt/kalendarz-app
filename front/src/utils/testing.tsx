import React, { ComponentType } from 'react';
import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { RootState, rootReducer } from '@store/index';

export function matchMedia(): void {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query: string) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(),
            removeListener: jest.fn(),
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });
}
export function isTestingEnv(): boolean {
    return process.env.JEST_WORKER_ID !== undefined || process.env.NODE_ENV === 'test';
}

export interface CustomOptions extends RenderOptions {
    preloadedState: RootState;
}

function render(
    ui: React.ReactElement,
    options?: CustomOptions,
) {
    const store = configureStore({ reducer: rootReducer, preloadedState: options?.preloadedState });
    function Wrapper({ children }: {children: React.ReactNode}) {
        return <Provider store={store}>{children}</Provider>;
    }

    return { element: rtlRender(ui, { wrapper: Wrapper as ComponentType, ...options }), store };
}

// re-export everything
export * from '@testing-library/react';
// override render method
export { render };
