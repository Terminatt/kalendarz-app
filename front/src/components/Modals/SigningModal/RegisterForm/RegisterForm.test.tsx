import { store } from '@store/index';
import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { fireEvent, matchMedia, render, waitFor } from '@utils/testing';
import RegisterForm from './RegisterForm';
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { userMock } from '@entity-mocks/User';

export const handlers = [
    rest.post('/users', (req, res, ctx) => {
      return res(ctx.json(userMock), ctx.delay(150))
    })
  ];

const server = setupServer(...handlers)

describe('Register Form Component', () => {
    beforeEach(() => {
        matchMedia();
    });
    beforeAll(() => server.listen());

    afterEach(() => server.resetHandlers());

    afterAll(() => server.close());

    it('matches the snapshot', () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <RegisterForm />
                </Provider>
            )
            .toJSON();
        
        expect(tree).toMatchSnapshot();
    });

    it('should call onFinishCallback', async () => {
        const onFinishCallback = jest.fn();
        const { element } = render(<RegisterForm 
            onFinishCallback={onFinishCallback} 
            initialValues={{...userMock, password: '123asdw2a!@', repeatPassword: '123asdw2a!@'}} 
        />);
        
        const btn = element.getByText('Zarejestruj się');
        fireEvent.click(btn)
        await waitFor(() => {
            expect(onFinishCallback).toBeCalledTimes(1);
        })
    });

    it('should not call onFinishCallback', async () => {
        const onFinishCallback = jest.fn();
        const { element } = render(<RegisterForm 
            onFinishCallback={onFinishCallback} 
        />);
        
        const btn = element.getByText('Zarejestruj się');
        fireEvent.click(btn)
        await waitFor(() => {
            expect(onFinishCallback).not.toBeCalled();
        })
    });

    it('appear an error with special character password', async () => {
        const { element } = render(<RegisterForm 
        />);
        
        const btn = element.getByText('Zarejestruj się');
        fireEvent.click(btn)
        await waitFor(() => {
            expect(element.getByText('Hasło musi zawierać chociaż jeden ze specjalnych symboli !, @, #, $, %, ^, &, *')).not.toBeNull();
        })
    });

    it('appear an error with numeric password', async () => {
        const { element } = render(<RegisterForm 
        />);
        
        const btn = element.getByText('Zarejestruj się');
        fireEvent.click(btn)
        await waitFor(() => {
            expect(element.getByText('Hasło nie może składać się z samych cyfr')).not.toBeNull();
        })
    });

});
