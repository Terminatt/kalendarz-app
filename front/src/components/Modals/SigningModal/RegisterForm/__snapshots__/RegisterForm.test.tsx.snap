// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Form Component matches the snapshot 1`] = `
<form
  className="ant-form ant-form-horizontal custom-form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ant-row ant-form-item half-input"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="firstName"
        title="Podaj imię"
      >
        Podaj imię
      </label>
    </div>
    <div
      className="ant-col ant-col-24 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="firstName"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Imię"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item half-input"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="lastName"
        title="Podaj nazwisko"
      >
        Podaj nazwisko
      </label>
    </div>
    <div
      className="ant-col ant-col-24 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="lastName"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Nazwisko"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="username"
        title="Podaj nazwę użytkownika"
      >
        Podaj nazwę użytkownika
      </label>
    </div>
    <div
      className="ant-col ant-col-24 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="username"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Nazwę użytkownika"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="email"
        title="Podaj email"
      >
        Podaj email
      </label>
    </div>
    <div
      className="ant-col ant-col-24 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="email"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Email"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="password"
        title="Podaj hasło"
      >
        Podaj hasło
      </label>
    </div>
    <div
      className="ant-col ant-col-24 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="password"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Hasło"
            type="password"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="ant-row ant-form-item"
    style={Object {}}
  >
    <div
      className="ant-col ant-col-24 ant-form-item-label"
      style={Object {}}
    >
      <label
        className="ant-form-item-required"
        htmlFor="repeatPassword"
        title="Powtórz hasło"
      >
        Powtórz hasło
      </label>
    </div>
    <div
      className="ant-col ant-col-24 ant-form-item-control"
      style={Object {}}
    >
      <div
        className="ant-form-item-control-input"
      >
        <div
          className="ant-form-item-control-input-content"
        >
          <input
            className="ant-input"
            id="repeatPassword"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Powtórz hasło"
            type="password"
            value=""
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="custom-form-btns"
  >
    <button
      className="ant-btn custom-btn custom-btn-primary custom-form-btn"
      disabled={false}
      onClick={[Function]}
      type="submit"
    >
      <span>
        Zarejestruj się
      </span>
    </button>
    <button
      className="ant-btn custom-btn custom-btn-clear custom-form-btn"
      onClick={[Function]}
      type="button"
    >
      <span>
        Wyczyść
      </span>
    </button>
  </div>
</form>
`;
